src/analysis/calculus/inverse.lean

```diff
@@ -0,0 +1,582 @@
+
+variables (f f' f'inv)
+
+/-- A local homeomorphism between E` and `f'.ker × F` sending level surfaces of `f`
+to horizontal subspaces. -/
+def implicit_to_local_homeomorph : local_homeomorph E (F × f'.ker) :=
+(hf.implicit_aux_has_fderiv hf').to_local_homeomorph _
+
+/-- Implicit function `g` defined by `f (g y x) = y`. -/
+def implicit_function : F → f'.ker → E :=
+function.curry $ (hf.implicit_aux_has_fderiv hf').local_inverse _ _ _
+
+end defs
+
+@[simp] lemma implicit_to_local_homeomorph_fst (x : E) :
+  ((hf.implicit_to_local_homeomorph f f' f'inv  hf').to_fun x).fst = f x :=
+rfl
+
+@[simp] lemma implicit_to_local_homeomorph_ker_snd (x : f'.ker) :
+  ((hf.implicit_to_local_homeomorph f f' f'inv hf').to_fun (x + a)).snd = x :=
+by simpa only [implicit_to_local_homeomorph, to_local_homeomorph_to_fun, add_sub_cancel]
+  using (continuous_linear_map.proj_ker_of_right_inverse_apply_idem _ _ hf' x)
+
+@[simp] lemma implicit_to_local_homeomorph_self :
+  (hf.implicit_to_local_homeomorph f f' f'inv hf').to_fun a = (f a, 0) :=
+prod.ext rfl $ by simpa using hf.implicit_to_local_homeomorph_ker_snd hf' 0
+
+lemma mem_implicit_to_local_homeomorph_source :
+  a ∈ (hf.implicit_to_local_homeomorph f f' f'inv hf').source :=
+mem_to_local_homeomorph_source _
+
+lemma mem_implicit_to_local_homeomorph_target :
+  (f a, (0 : f'.ker)) ∈ (hf.implicit_to_local_homeomorph f f' f'inv hf').target :=
+by simpa only [implicit_to_local_homeomorph_self] using
+  ((hf.implicit_to_local_homeomorph f f' f'inv hf').map_source $
+    (hf.mem_implicit_to_local_homeomorph_source hf'))
+
+/-- `implicit_function` sends `(z, y)` to a point in `f ⁻¹' z`. -/
+lemma map_implicit_function_eq :
+  ∀ᶠ p in 𝓝 (f a, (0 : f'.ker)),
+    f (hf.implicit_function f f' f'inv hf' (p : F × f'.ker).1 p.2) = p.1 :=
+((hf.implicit_to_local_homeomorph f f' f'inv hf').eventually_right_inverse $
+  hf.mem_implicit_to_local_homeomorph_target hf').mono $ λ ⟨z, y⟩ h,
+    congr_arg prod.fst h
+
+/-- Any point in some neighborhood of `a` can be represented as `implicit_function`
+of some point. -/
+lemma eq_implicit_function :
+  ∀ᶠ x in 𝓝 a, hf.implicit_function f f' f'inv hf' (f x)
+    ((hf.implicit_to_local_homeomorph f f' f'inv hf').to_fun x).snd = x :=
+(hf.implicit_aux_has_fderiv hf').eventually_left_inverse
+
+/-- Derivative of the inverse function used to prove the implicit function theorem. -/
+lemma to_implicit_function_aux :
+  has_strict_fderiv_at (hf.implicit_to_local_homeomorph f f' f'inv hf').inv_fun
+    (f'inv.coprod $ subtype_val f'.ker) (f a, 0) :=
+hf.implicit_to_local_homeomorph_self hf' ▸
+  (hf.implicit_aux_has_fderiv hf').to_local_inverse
+
+lemma to_implicit_function :
+  has_strict_fderiv_at (hf.implicit_function f f' f'inv hf' (f a)) (subtype_val f'.ker) 0 :=
+begin
+  have := (hf.to_implicit_function_aux hf').comp 0
+    ((has_strict_fderiv_at_const (f a) 0).prod $ has_strict_fderiv_at_id 0),
+  convert this,
+  ext x,
+  simp
+end
+
+end has_strict_fderiv_at
+
```

Comment:
You could put a docstring with a section header here saying that you turn to the implicit function theorem.